### PR TITLE
Misc loadout builder work

### PR DESCRIPTION
**Pull request by kyleshay**
Another pass at loadout builder

This PR sets us up for some better filters down the line. We now build out base stats for items and compute their bonuses.

I've removed options for the loadout builder by moving things around.

Character selection now takes full use of the screen on desktop. 

rec3

### CODE REVIEW

**Comment by kyleshay** [comment]
updated shot.. still a work in progress. 
screen shot 2018-10-26 at 4 34 36 pm

File: src/app/character-select/CharacterSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+          >
+            <SimpleCharacterTile character={store} onClick={this.selectCharacter} />
+          </div>
+        ))}
+      </div>
+    );
+  }
+
+  private selectCharacter = (storeId: string) => {
+    this.setState({ dropdownOpen: false });
```
Comment (bhollis) [inline src/app/character-select/CharacterSelect.tsx]: What's this for?

File: src/app/d2-loadout-builder/LoadoutBuilder.tsx [hunk 2]
```diff
@@ -230,11 +216,20 @@ export class LoadoutBuilder extends React.Component<Props & UIViewInjectedProps,
           if (lockedItem.type === 'exclude') {
             filteredItems[bucket] = filteredItems[bucket].filter(
               (item) =>
-                !lockedMap[bucket].find((excludeItem) => excludeItem.item.index === item.index)
+                !lockedMap[bucket].find(
+                  (excludeItem) => (excludeItem.item as D2Item).index === item.index
```
Comment (bhollis) [inline src/app/d2-loadout-builder/LoadoutBuilder.tsx]: You shouldn't need to do `as D2Item`.

File: src/app/d2-loadout-builder/LoadoutBuilder.tsx [hunk 3]
```diff
@@ -347,66 +342,66 @@ export class LoadoutBuilder extends React.Component<Props & UIViewInjectedProps,
...
-          processRunning={processRunning}
-          processedSets={processedSets}
-          lockedMap={lockedMap}
-          selectedStore={selectedStore}
-          onLockChanged={this.updateLockedArmor}
-        />
+        )}
+        {(processedSets.length === 0 &&
+          this.state.requirePerks &&
+          !this.foundSets && (
```
Comment (bhollis) [inline src/app/d2-loadout-builder/LoadoutBuilder.tsx]: Use `this.foundSets ? (...) : (...)`

File: src/app/d2-loadout-builder/LoadoutBuilderItem.tsx [hunk 4]
```diff
@@ -27,7 +27,9 @@ export default function LoadoutBuilderItem({
       <ItemPopupTrigger item={item}>
         <div
           className={classNames({
-            'excluded-item': locked && locked.some((p) => p.item.index === item.index)
+            'excluded-item':
+              locked &&
+              locked.some((p) => p.type === 'exclude' && (p.item as D2Item).id === item.id)
```
Comment (bhollis) [inline src/app/d2-loadout-builder/LoadoutBuilderItem.tsx]: You don't need `as D2Item` to get an id.

File: src/app/d2-loadout-builder/generated-sets/utils.ts [hunk 5]
```diff
@@ -148,12 +207,6 @@ export function toggleLockedItem(
...
-export function isInventoryItemDefinition(
-  item: D2Item | DestinyInventoryItemDefinition
-): item is DestinyInventoryItemDefinition {
-  return Boolean((item as DestinyInventoryItemDefinition).displayProperties);
-}
-
-export function isD2Item(item: D2Item | DestinyInventoryItemDefinition): item is D2Item {
-  return Boolean((item as D2Item).name);
+export function isD2Item(item: LockedItemType['item']): item is D2Item {
+  return Boolean(item as D2Item);
```
Comment (bhollis) [inline src/app/d2-loadout-builder/generated-sets/utils.ts]: This will always return `true`. You gotta inspect the structure.

**Comment by kyleshay** [comment]
Sooo, pretty happy with where it is at now. Still need to style the autocomplete... that part is still in beta.

Here is what things currently look like:
rec3

**Comment by inexorableAce** [comment]
This is looking better for sure! I still feel like the perk locks need to be AND not OR. If I lock two perks I expect it to return what I locked and nothing else. Even if that means I don't have it. 

Ideally we could prevent that scenario by filtering the selectable perks as we make selections.

This is awesome though a lot more flexibility for the user in these enhancements!

**Comment by inexorableAce** [comment]
I feel like they all should be AND, just the term lock and the way it works for armor suggests that’s the way it is. 

It’s a bit confusing when I ‘lock’ a perk and it shows me things without what I locked. 

The soft fail is kinda nice though perhaps we have a section under the found loadouts for ‘close matches’ as a soft fail ‘you don’t have what you want, but these are close’ kinda thing. Best of both worlds.

**Comment by inexorableAce** [comment]
that's awesome, and since the list of builds updated immediately with the perk selection still open, it's not as much of an issue that perks don't get greyed out yet. I think it's good to push out and get another round of feedback on. From the discord comments, I think is gonna be a hit.

**Comment by kyleshay** [comment]
Glorious!

screen shot 2018-10-27 at 11 18 09 am

**Comment by kyleshay** [comment]
Thanks for the feedback @inexorableAce. It's feeling really good right now!! :D

I think that I'll do some final polishing to the auto complete and general styling in another PR... this one is already so large.

@bhollis think we're ok to merge? (it's still only set to beta users)

**Pull request merged**